add QNN tutorial

MeltemTolunay opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
QNN training tutorial with QCNN ansatz


### Details and comments

miamico commented:
Very quick comment, let's move the notebook to the `docs/notebooks` folder

miamico commented:
Other comments to address to make this in shape for a PR:
- Try adding SPSA instead of COBYLA. The vanilla implementation is quite contained so it could be included in the notebook
- Recast the notebook into the Qiskit Patterns framework (1. specify the classical problem and the quantum way to solve it, 2. optimize it for the hardware 3. execute, 4.post-process)
- Add more details explaining what is happening in each cell, right now the explanations are very bare-bone

MeltemTolunay commented:
I now added a gradient function with parameter shift rule, qiskit patterns formalism, code to run on hardware and more text explaining the code.

MeltemTolunay commented:
New changes include:
- ParameterVectorElements are ordered alphabetically. FeatureMap parameters now have a new parameter_prefix='a' to ensure that input and weight parameters are assigned in correct order.
- Parameter shift rule for computing the gradients are verified.
- Added a new function to compute the gradient of the MSE loss with respect to weight parameters.
- Added new optimizer options. We can now select from a collection of gradient-free optimizers, gradient-based optimizers and SPSA.
In addition, Mirko's previous comments regarding SPSA, Qiskit Patterns framework, and cell explanations are fully addressed.
